Re-write to make more reliable

oversword opened the pull request:
**Issue:** Having two light levels in the same position can cause flickering between them
**Expected behaviour:** ideally the highest light level should be displayed
**Fix:** tracking multiple items by position, calculating the correct level if an update is required

**Issue:** Using the fast priv + fly priv can cause lights to be left behind in your path
**Expected behaviour:** Light nodes should be removed no matter how far away from them you are
**Fix:** Use the global step to run updates on all nodes instead of the node timer

**Issue:** It's not easy to add in a new custom dynamic light
**Expected behaviour:** there should be an API with plenty of helpful methods for getting data and registering new lights
**Fix:** Many new helper methods added that were extracted from the original code, as well as new methods added for tracking new wielded lights and running code at the same time as the update


### API:

- `wielded_light.register_lightstep(callback)`
  Registers a callback to be called every time the update interval is passed
- `wielded_light.register_player_lightstep(callback)`
  Registers a callback to be called for each player every time the update interval is passed
- `wielded_light.wielded_light_of_level(light_level)`
  returns the node name for a given light level
- `wielded_light.level_of_wielded_light(node_name)`
  gets the light level for a given node name, inverse of `wielded_light_of_level`
- `wielded_light.is_wielded_light(node_name)`
  check if a node name is one of the wielded light nodes
- `wielded_light.is_lightable_node(node_pos)`
  check if node can have a wielded light node placed in it, currently only air and other wielded nodes
- `wielded_light.get_light_position(pos)`
  gets the closest position to pos that's a lightable node
- `wielded_light.get_light_level(item_name)`
  gets the emitted light level of a given item name
- `wielded_light.register_item_light(itemname, light_level)`
  already exists - register an item as shining
- `wielded_light.track_item_entity(userdata, cat, item)`
  Keep track of an item entity, should be called once, will update the position of the entity automatically but assumes the light level is constant
- `wielded_light.track_user_entity(userdata, cat, item)`
  Keep track of a user / player entity, should be called as often as the user updates, will update the position of the entity automatically and expects you to provide a different wielded item each time


### Efficiency concerns:
I've tried to make sure this doesn't take much time to run, the nodes are updated as little as possible and I've opted for increased memory usage to reduce the operations that need to be done.
Two tables are maintained: a table of entities, their positions, and the items associated with them; and a table of positions with the items and light levels stored.
Each step every entity is run through to update its position, if changes have occurred then the light is recalculated for that position.

oversword commented:
Can take another pass at this once https://github.com/minetest-mods/wielded_light/pull/10 is merged

oversword commented:
@bell07 thoughts?

bell07 commented:
From my point of view the #10 can be merged to the master.

This PR #11 contains a lot of changes. Since I do not play & hack minetest anymore, I cannot decide if it's fine or bad.

@SmallJoker , you take on the under water shine PR. Maybe you can look into this PR too?

SmallJoker commented on init.lua:
```diff
@@ -1,77 +1,302 @@
+local cleanup_interval = update_interval*3
+local velocity_projection = update_interval * 1
 local level_delta = 2
+local max_light_level = 14
```
Use `minetest.LIGHT_MAX`

oversword replied:
That's done, it will change behaviour on any server where this is not 14 though - as this was previously hard coded as 14

SmallJoker commented on init.lua:
```diff
@@ -1,77 +1,302 @@
+local mod_name = "wielded_light"
```
If mod.conf is set up properly, use `minetest.get_current_modname()`

oversword replied:
Done

SmallJoker commented on init.lua:
```diff
@@ -1,77 +1,302 @@
+	local light_level = wielded_light.get_light_level(item)
+	if light_level <= 0 then return end
+
+	local uid = tostring(userdata)
```
What does this do? How can it convert `ObjectRef` to anything usable?

oversword replied:
It converts the object to something like `"userdata: 0x41836ab0"`, which should be suitable as a unique ID

SmallJoker commented on init.lua:
```diff
@@ -1,77 +1,302 @@
+	local tracked_entity = tracked_entities[uid]
+	local tracked_item = tracked_entity.items[id]
+
+	if not tracked_item
+	or tracked_item.item ~= item
+	then
```
Weird indent level. A single line for all of them would be fine.

oversword replied:
It's easier for commenting out one part of the condition at a time, this one was probably lot bigger before it was simplified

SmallJoker commented on init.lua:
```diff
@@ -1,77 +1,302 @@
+	table.insert(update_player_callbacks, callback)
+end
+
+function wielded_light.wielded_light_of_level(light_level)
```
Needs naming improvement. `get_light_node_name` or similar.

oversword replied:
Naming it like this makes it clearer that it's the inverse of `level_of_wielded_light`

SmallJoker commented on init.lua:
```diff
@@ -1,77 +1,302 @@
 local update_interval = 0.2
+local cleanup_interval = update_interval*3
+local velocity_projection = update_interval * 1
 local level_delta = 2
```
What does this do? Please add a short, helpful comment to the code.

oversword replied:
Which bit? `level_delta`? I didn't even add that...

SmallJoker replied:
I think `velocity_projection` and `level_delta` would both benefit from a short comment. I do know that `level_delta` was not changed, but if you happen to know what it does, it would be a great opportunity to document it a bit.

oversword replied:
I've added in descriptions for all the constants at the top here

SmallJoker commented on init.lua:
```diff
@@ -1,77 +1,302 @@
+	return entity and (entity.obj:is_player() or entity.obj:get_entity_name() or false)
+end
+
+local function entity_pos(userdata, offset)
```
You could also call `userdata` -> `obj` or `entity` to clarify that it's an ObjectRef. Same for the other functions, where applicable.

oversword replied:
Yeah I've renamed that to `obj`, I think in this code we should be using `entity` to refer to the custom object made for tracking and `obj` for the actual userdata / ObjectRef being tracked

SmallJoker commented on init.lua:
```diff
@@ -1,77 +1,302 @@
-					do_update = true
+local cat_id = 0
+local cat_codes = {}
+local function get_cat_code(cat)
```
Rename suggestion: `get_light_category_id()`.

More importantly: Why do you need categories?
Whatever ID it outputs - it's not reliable between server restarts when a mod is enabled to adds (or makes use of) another category.

oversword replied:
It doesn't need to be reliable over restarts, the entities being referred to will also change their IDs so all the old lights should get removed and new ones created.

Categories are required so you can assign multiple wielded lights to the same entity without getting them confused - this is because of a new item I am trying to add that acts like a "mining helmet", basically a helmet that glows so you can mine without placing torches down or requiring an "off-hand".
This concept brings in the the question of what happens when you try to impose two wielded light values (on the head and in the hand) on the same space in different globalstep loops - it flickers between them.
Having categories allows you to keep any number (..well, up to 256.. any number within reason) of wielded light items on the same entity at a time, if you wanted a full set of glow armor, a torch in one hand and glow glass in the other, you could have a set of categories, one for each armor position and one for each hand. This makes it possible to keep track of what light source is where when otherwise they would get lost in whatever process you are using to sum up the light levels, making the light level unreliable.

oversword replied:
Renamed

SmallJoker replied:
Excuse me, but why couldn't we just use the category name directly? That's already an unique identifier for the type of light.

oversword replied:
I did add a comment at the top of this function that explains that

oversword replied:
```
	Using 2-digit hex codes for categories
	Starts at 00, ends at FF
	This makes it easier extract `uid` from `cat_id..uid` by slicing off 2 characters
	The category ID must be of a fixed length (2 characters)
```

SmallJoker commented on init.lua:
```diff
@@ -1,77 +1,302 @@
+				if item.level > 0 then
+					add_light(pos_str, id, item.level)
+				else
+					remove_light(pos_str, id)
```
When does this case occur?

oversword replied:
When the user switches from holding an item with a light level to holding an item with no light level

SmallJoker commented on init.lua:
```diff
@@ -1,77 +1,302 @@
+local function update_entity(entity)
+	local pos = entity_pos(entity.obj, entity.offset)
+	local pos_str = pos and minetest.pos_to_string(pos)
+	if entity.pos and pos_str ~= entity.pos then
```
Could you please add a few comments describing why and what's happening here.

oversword replied:
@SmallJoker Is that clearer?

SmallJoker replied:
Yes! Thank you very much.

SmallJoker commented on init.lua:
```diff
@@ -1,77 +1,302 @@
+	return cat_codes[cat]
+end
+
+local function entity_still_exists(entity)
```
Rename suggestion: `is_entity_valid`

oversword replied:
Are you sure? That seems less descriptive of what it's checking for

SmallJoker replied:
It's supposed to clarify that the function is a check in first place. The part `still` is superfluous/misleading, because the functions' purpose is to check whether an entity is valid, and not whether something has changed since the last call.

oversword replied:
changed now

SmallJoker reviewed (commented):
I don't understand much of what's going on here, but would like to see a few more comments to ease comprehension of the code.
I also wonder whether this much code is required at all. You could create a node position-indexed table and keep the `ObjectRef` references. On timer elapse you could then check whether at least 1 reference is still valid, otherwise remove. However, it's not recommended to keep ObjectRefs outside of their callbacks; and to avoid that would basically make this PR impossible from what I can see.

SmallJoker commented:
Whereas the code is in proper shape in terms of styling, I am very skeptic whether it's the correct approach.

oversword commented:
@SmallJoker 
> I don't understand much of what's going on here, but would like to see a few more comments to ease comprehension of the code.
Fully commented now, let me know if you're still having trouble understanding it

> I also wonder whether this much code is required at all. You could create a node position-indexed table and keep the `ObjectRef` references. On timer elapse you could then check whether at least 1 reference is still valid, otherwise remove. However
Most of the code is there to stop the rest of it running too often - see "Efficiency Concerns". The simplification you proposed would make it much less efficient, the way it has been written avoids re-calculation as often as possible and tries to minimise the amount of work that's actually done on the global step.

> it's not recommended to keep ObjectRefs outside of their callbacks; and to avoid that would basically make this PR impossible from what I can see.
Recommended by who? I see no other way to keep track of their position. The way this mod works currently means stale light sources can be left all over the place long after the entity that produced them has gone, we need to track ownership of these light sources so we can remove them more reliably. See the three issues detailed at the top of my PR.

SmallJoker commented:
> The simplification you proposed would make it much less efficient, the way it has been written avoids re-calculation as often as possible and tries to minimise the amount of work that's actually done on the global step.

Yes, that's also what I understood from the goals of this PR. I however wasn't sure whether there wasn't any simpler, yet efficient approach. For that I'd need to fully understand the code; which is currently not the case as you might've guessed by some of my questions.

> Recommended by who? I see no other way to keep track of their position.

Older Minetest versions had strange side-effects when entities were used after their lifecycle. However, at least since last year this should be sorted out. Functions will just return `nil` when the object turned invalid. Please ignore that comment, it's outdated.

oversword commented:
>  I however wasn't sure whether there wasn't any simpler, yet efficient approach.

If you can think of one please let me know! This code is my third draft, previously it was even more complex, this is as simple as I can conceive it, and I have tried other methods of making it simpler but they all result in some issue.
The problem is multiplicity, I'm trying to keep track of multiple positions with multiple entities holding multiple light sources, the code I've written should be capable of maintaining reliable behaviour in indefinitely complex scenarios - which you might not expect, but are totally possible - see the various implementations of trampolines to really stretch the physics engine!

> For that I'd need to fully understand the code; which is currently not the case as you might've guessed by some of my questions.

Tell me if I'm wrong, but I think the main point of confusion may be the two tables `active_lights` and `tracked_entities` why do we need both?
`tracked_entities` is indexed by entity, this means when the update loops runs through all entities and recalculates their position it only has to do so once per entity - which would be more complex otherwise
`active_lights` is indexed by position, this makes it better for recalculating light for a node and for checking if a node still has any light emitted in it (see the cleanup function)


> Older Minetest versions had strange side-effects when entities were used after their lifecycle. However, at least since last year this should be sorted out. Functions will just return `nil` when the object turned invalid. Please ignore that comment, it's outdated.

I did notice that actually, the `is_entity_valid` function depends on that behaviour, frankly if they wanted to avoid side effects they should be returning the objects by copy, not reference, it would still be possible to identify them with other means such as players by their username - but it would make what I've written even more convoluted.

SmallJoker commented:
I yet don't know why, but with this PR, the light updates become much more apparent. Sometimes entire mapblocks flicker while moving. However, in `master` this does only happen rarely whereas it's still moving the player's light at the same speed.

Configuration:
  * Server hosting
  * `dedicated_server_step = 0.08`

oversword commented on init.lua:
```diff
@@ -1,77 +1,411 @@
+function remove_light(pos, id)
+	if not active_lights[pos] then return end
+	-- minetest.log("error", "rem "..id.." "..pos)
+	-- minetest.after(update_interval, function ()
+	active_lights[pos][id] = nil
+	light_recalcs[pos] = true
+	-- end)
```
@SmallJoker

> Sometimes entire mapblocks flicker while moving.

Thanks for testing, I thought I might be the only one seeing that. Uncommneting this delay fixes it for me.
The two lists are decoupled enough that delaying it should have no side effects, I just didn't want to leave this in without knowing it was an issue as it seems a little more random than other know causes of flickering.
Basically, the light doesn't seem to update immediately, leaving the last light in for another round seems to smooth that out - some flickering still noticeable around the edges if you look closely, but I think that's unavoidable.

I'll fix this now.

oversword replied:
Actually I just thought of  way to break this, let me think about it for a sec

oversword replied:
Alright give that a shot, pretty sure it can't break and fixes the flickering

SmallJoker replied:
How could this cause bugs? Is the light information missing in the timer/globalstep whenever the position changes?

Let me know when you're done; and I'll give it a try again.

oversword replied:
It's done, that's what I meant with the last comment
> Alright give that a shot, pretty sure it can't break and fixes the flickering

So, originally I had
```
minetest.after(update_interval, function ()
	active_lights[pos][id] = nil
	light_recalcs[pos] = true
end)
```
which I noticed may cause an issue if I move away from and then back to the same spot within one update_interval. When I get back to the spot it may erroneously remove the light source as it had been marked for removal in the future.

The code now contains (something like) this:
```
active_lights[pos][id] = nil
minetest.after(update_interval, function ()
	light_recalcs[pos] = true
end)
```

Which fixes it in pretty much the same way. It updates the list synchronously, but doesn't ask for a synchronous update, it ask for it on the next update step. This means that the list will always be correct, but the light in the old position will hang around for another step before the position is updated.

SmallJoker reviewed (approved):
Works.

bell07 commented:
Does this changes work under water too?

oversword commented:
@bell07 not yet but it shouldn't be hard to merge with the current underwater branch... so long as it hasn't been merged with master yet

SmallJoker commented:
@bell07  Is the under water shine branch ready for merge? If you're too busy, I would merge it into `master` and attempt to solve conflicts, if there are any. Otherwise this PR will wait for an eternity until something happens.

bell07 commented:
@SmallJoker, I am away from Minetest, so I did not tried the last changes. Currently I do not have minetest installed on my PC.  Looked into code the changes seems to be fine. Please take over and merge the changes into master, then delete the underwater-branch.

SmallJoker commented:
:+1: under water feature integrated.

oversword commented:
Thank you @SmallJoker will have the conflicts resolved by this time tomorrow

oversword commented:
Took a little longer than expected - have been messing around with the underwater light feature as I'm 100% happy with it, I will probably try and get it to work with flowing water if possible, if not it has the same functionality as it currently does in this PR

Updates since last time:
- Made a generic registration function for lightable nodes
- Added the ability to provide custom "prefixes" to keep the existing node names the same, otherwise the prefix will be generated from the registered node name. The prefixes are checked for uniqueness
- Registered the floodable flag in addition to the light level in `shiny_items`, made the registration function a bit more complex to account for different registration scenarios
- Renamed `level_of_wielded_light` and the other one to `level_of_lighting_node` as per new registration index

SmallJoker commented:
Hmm. The PR works, except that I somehow cannot get the light to work underwater (water source node). Only `air` works so far.

oversword commented:
@SmallJoker What light are you using? Make sure it's not a torch as that is floodable

SmallJoker commented:
Ah yes. I was using the regular torch which (for whatever reason) worked partially on `master`, hence I consider this as a reasonable bugfix.

Will merge in a few days unless anyone has an objection.

oversword commented:
> I was using the regular torch which (for whatever reason) worked partially on `master`

Yeah so there may be a change of behavior there, previously it (may) have seen the water source as an invalid position and picked the air above it sometimes, in this version it considers the water source valid as a position but rejects it further down the line for being floodable + liquid, could probably resolve that if you need it but I think this is reasonable behavior and possibly a bugfix as you say

oversword commented:
:partying_face:
